### ISSUE TITLE
DAG API is for IPLD only

### ISSUE THREAD

**Issue by vmx** [ipfs/ipfs#326]
The DAG API in IPFS is for IPLD. Hence the traversing/addressing of nodes should be like they are represented in IPLD.

At the moment when using the DAG API in go-ipfs and you traverse dag-pb (aka. Protocol Buffer Nodes or UnixFS Nodes) then you address other nodes directly with their CID. You also can't e.g. access the `Data` field. This behaviour is wrong. The correct behaviour is to work on the IPLD level, where there is the `Data` field and the linked nodes are a list of nodes within the `Links` field.

This change needs to happen in several places, this is the tracking issue for it.

DAG API works on IPLD-level:

 - [ ] go-ipfs
 - [x] js-ipfs

An open question is, what should happen if you pass in an IPFS/IPNS namespaced path (e.g. `/ipfs/<cid>`)?

IPFS/IPNS namespaced path are handled correctly:

 - [ ] go-ipfs
 - [ ] js-ipfs

/cc @Stebalien @diasdavid @alanshaw

**Comment by vmx** [comment]
> An open question is, what should happen if you pass in an IPFS/IPNS namespaced path (e.g. /ipfs/<cid>)?

I think there should be an error message that the DAG API only works for IPLD and you should use the Files API instead.

**Comment by magik6k** [comment]
In go-ipfs there is a PR partially addressing this - https://github.com/ipfs/go-ipfs/pull/4672 and some related discussion - https://github.com/ipfs/go-ipfs/issues/4666.

tl;dr is that `/ipfs` paths by default resolve unixfs/hamts/etc, and `/ipld` resolves raw ipld paths (so you get the access to `Data` and `Links` nodes of dag-pb). Paths without prefix are forbidden (at least at the api level, cli commands will likely prepend default prefix depending on which command is used).

Dag commands will use `/ipld` by default

**Comment by vmx** [comment]
@magik6k I've updated the issue and my comment. It's not really an open question (sorry for the confusion). The DAG API should error if an IPFS/IPNS path is passed in.
